File: module_5/option_b/exercise_1/solution_merge.py
"""
Session 2 Solution — Part 1: Merge and Explore
Merges the survey data with the demographics file using a left join.
"""

import pandas as pd
import numpy as np

# ── Load data ────────────────────────────────────────────────────
survey = pd.read_csv("module_5/option_b/data/swiss_franc_survey.csv")
demog = pd.read_csv("module_5/option_b/data/respondent_demographics.csv")

print("Survey shape:", survey.shape)
print("Demographics shape:", demog.shape)
print(f"Survey unique IDs: {survey['respondent_id'].nunique()}")
print(f"Demographics unique IDs: {demog['respondent_id'].nunique()}")

# ── Merge: left join on respondent_id ────────────────────────────
merged = survey.merge(demog, on="respondent_id", how="left")

# ── Verify the merge ─────────────────────────────────────────────
print("\n=== MERGE VERIFICATION ===")
print(f"Merged shape: {merged.shape}")
assert merged.shape[0] == 2044, f"Expected 2044 rows, got {merged.shape[0]}"
print("Row count: 2,044 (no rows lost or gained)")

assert merged["respondent_id"].nunique() == 2044, "Duplicate respondent IDs!"
print("No duplicate respondent IDs")

# ── Explore the data ─────────────────────────────────────────────
print("\n=== TREATMENT GROUPS ===")
print(merged["treatment"].value_counts().sort_index())

print("\n=== OUTCOME DISTRIBUTION ===")
print(merged["supports_intervention"].value_counts(dropna=False))
print(f"\nUnweighted proportion supporting: {merged['supports_intervention'].mean():.4f}")

weighted_prop = np.average(
    merged["supports_intervention"].dropna(),
    weights=merged.loc[merged["supports_intervention"].notna(), "survey_weight"],
)
print(f"Weighted proportion supporting: {weighted_prop:.4f}")

print("\n=== FX EXPOSURE ===")
print(merged["fx_status"].value_counts(dropna=False))

print("\n=== MISSING DATA ===")
for col in merged.columns:
    n_miss = merged[col].isna().sum()
    if n_miss > 0:
        print(f"  {col}: {n_miss} missing ({n_miss/len(merged)*100:.1f}%)")

print("\n=== INCOME QUINTILE = 0 ===")
print(f"Respondents with income_quintile = 0: {(merged['income_quintile'] == 0).sum()}")

# ── Save ─────────────────────────────────────────────────────────
merged.to_csv("module_3/output/merged_survey.csv", index=False)
print("\nSaved merged data to module_3/output/merged_survey.csv")

File: module_5/option_b/exercise_1/solution_merge.R
# Session 2 Solution — Part 1: Merge and Explore (R version)
# Merges the survey data with the demographics file using a left join.

library(dplyr)
library(readr)

# ── Load data ────────────────────────────────────────────────────
survey <- read_csv("module_5/option_b/data/swiss_franc_survey.csv")
demog  <- read_csv("module_5/option_b/data/respondent_demographics.csv")

cat("Survey shape:", dim(survey), "\n")
cat("Demographics shape:", dim(demog), "\n")
cat("Survey unique IDs:", length(unique(survey$respondent_id)), "\n")
cat("Demographics unique IDs:", length(unique(demog$respondent_id)), "\n")

# ── Merge: left join on respondent_id ────────────────────────────
merged <- survey %>% left_join(demog, by = "respondent_id")

# ── Verify the merge ─────────────────────────────────────────────
cat("\n=== MERGE VERIFICATION ===\n")
cat("Merged shape:", dim(merged), "\n")
stopifnot(nrow(merged) == 2044)
cat("Row count: 2,044 (no rows lost or gained)\n")

stopifnot(length(unique(merged$respondent_id)) == 2044)
cat("No duplicate respondent IDs\n")

# ── Explore the data ─────────────────────────────────────────────
cat("\n=== TREATMENT GROUPS ===\n")
print(table(merged$treatment))

cat("\n=== OUTCOME DISTRIBUTION ===\n")
print(table(merged$supports_intervention, useNA = "always"))
cat("\nUnweighted proportion supporting:",
    round(mean(merged$supports_intervention, na.rm = TRUE), 4), "\n")

non_na <- !is.na(merged$supports_intervention)
weighted_prop <- weighted.mean(
  merged$supports_intervention[non_na],
  w = merged$survey_weight[non_na]
)
cat("Weighted proportion supporting:", round(weighted_prop, 4), "\n")

cat("\n=== FX EXPOSURE ===\n")
print(table(merged$fx_status, useNA = "always"))

cat("\n=== MISSING DATA ===\n")
for (col in names(merged)) {
  n_miss <- sum(is.na(merged[[col]]))
  if (n_miss > 0) {
    cat(sprintf("  %s: %d missing (%.1f%%)\n",
                col, n_miss, n_miss / nrow(merged) * 100))
  }
}

cat("\n=== INCOME QUINTILE = 0 ===\n")
cat("Respondents with income_quintile = 0:",
    sum(merged$income_quintile == 0, na.rm = TRUE), "\n")

# ── Save ─────────────────────────────────────────────────────────
write_csv(merged, "module_5/option_b/output/merged_survey.csv")
cat("\nSaved merged data to module_5/option_b/output/merged_survey.csv\n")
